### PR TITLE
feat(factor): recovery code factor plugin implemented

### PR DESCRIPTION
**Pull request by ashraf706**
A new factor plugin implemented so that user can generate and download recovery code.
This plugin properties are configurable. A thyme leaf template is also added in forms section for this plugin.

This pull request relates to : https://github.com/gravitee-io/issues/issues/7014

### CODE REVIEW

**Review by remisultan — CHANGES_REQUESTED** [review]
A few comments.

Also having a postman series of test would strengthen the feature

File: gravitee-am-factor/gravitee-am-factor-recovery-code/src/main/java/io/gravitee/am/factor/recovery/code/provider/RecoveryCodeFactorProvider.java [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+     * This function use apache commons RandomStringUtils to generate alphanumeric random string
+     * @return list of random strings
+     */
+    private List<String> recoveryCodes() {
+        final int minMaxLength = configuration.getDigit();
+        logger.debug("Generating recovery code of {} digits", minMaxLength);
+
+        ArrayList<String> recoveryCodes = new ArrayList<>();
+        for (int i = 0; i < configuration.getCount(); i++) {
+            recoveryCodes.add(RandomStringUtils.randomAlphanumeric(minMaxLength, minMaxLength));
```
[the remark this answers is not in the document]
Comment (ashraf706) [reply gravitee-am-factor/gravitee-am-factor-recovery-code/src/main/java/io/gravitee/am/factor/recovery/code/provider/RecoveryCodeFactorProvider.java]: Updated as suggested

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-account/src/main/java/io/gravitee/am/gateway/handler/account/resources/AccountFactorsEndpointHandler.java [hunk 2]
```diff
@@ -184,6 +188,17 @@ public void enrollFactor(RoutingContext routingContext) {
                     }
                 }
 
+                //Check if the facto is a recovery factor.
```
[the remark this answers is not in the document]
Comment (ashraf706) [reply gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-account/src/main/java/io/gravitee/am/gateway/handler/account/resources/AccountFactorsEndpointHandler.java]: typo fixed

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-account/src/main/java/io/gravitee/am/gateway/handler/account/resources/AccountFactorsEndpointHandler.java [hunk 3]
```diff
@@ -213,6 +228,20 @@ public void enrollFactor(RoutingContext routingContext) {
         }
     }
 
+    private Completable generateRecoveryCode(RoutingContext routingContext, Factor factor, FactorProvider factorProvider) {
```
[the remark this answers is not in the document]
Comment (ashraf706) [reply gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-account/src/main/java/io/gravitee/am/gateway/handler/account/resources/AccountFactorsEndpointHandler.java]: 3rd parameter FactorProvider changed to RecoveryFactor. The second parameter Factor is actually data model.

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/RootProvider.java [hunk 4]
```diff
@@ -505,6 +513,7 @@ private void csrfHandler(Router router) {
         router.route(PATH_LOGIN_SSO_POST).handler(csrfHandler);
         router.route(PATH_MFA_CHALLENGE).handler(csrfHandler);
         router.route(PATH_MFA_CHALLENGE_ALTERNATIVES).handler(csrfHandler);
+        router.route(PATH_MFA_RECOVERY_CODE).handler(csrfHandler);
```
Comment (leleueri) [inline gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/RootProvider.java:516]: Do we need to set the cspHandler for RecoveryCode route ? (line 531)

File: gravitee-am-factor/gravitee-am-factor-recovery-code/src/main/java/io/gravitee/am/factor/recovery/code/provider/RecoveryCodeFactorProvider.java [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+        final List<String> recoveryCodes = (List<String>) enrolledFactor.getSecurity().getAdditionalData().get(RECOVERY_CODE);
+
+        return Completable.create(emitter -> {
+            if (recoveryCodes.contains(code)) {
+                //remove the code from the list as the recovery is not re-usable
+                recoveryCodes.remove(code);
+                enrolledFactor.getSecurity().setAdditionalData(Map.of(RECOVERY_CODE, recoveryCodes));
+                emitter.onComplete();
+            } else {
+                logger.error("An error occurred while validating with recovery code");
```
Comment (leleueri) [inline gravitee-am-factor/gravitee-am-factor-recovery-code/src/main/java/io/gravitee/am/factor/recovery/code/provider/RecoveryCodeFactorProvider.java:72]: Don't think it is something we should log as ERROR. An invalid recovery code is a standard case.
IMHO, it is at most a WARN but INFO or DEBUG may be fine.

**Review by leleueri — CHANGES_REQUESTED** [review]
In addition of the two comments, here is 2 test cases on which I wondering if we should do something.

* First case
1 - enable a factor (SMS or OTP) on application and recovery code 
2 - login a user and register the factor
3 - validate the factor challenge
4 - revory code page is displayed. 
5 - do not click on the button but close the browser (or close the session)
6 - login the user again and validate the factor challenge
7 - the recovery code generation isn't done as it has been done during the first connection but as the session has been closed, but as we didn't click on the button can we consider that the user copied the code ?

* Second case:

1 - enable a factor (SMS or OTP) on application but NOT the recoevry code
2 - login a user and register the factor
3 - validate the factor challenge
4 - logout the user
5 - enable the recovery code on the app
6 - login the user
7 - validate the factor challenge
8 - recovery code are not displayed. (I expected recovery code page with my codes as the option has been enabled)

File: gravitee-am-factor/gravitee-am-factor-recovery-code/src/main/java/io/gravitee/am/factor/recovery/code/provider/RecoveryCodeFactorProvider.java (hunk 5 above)
Comment (ashraf706) [reply gravitee-am-factor/gravitee-am-factor-recovery-code/src/main/java/io/gravitee/am/factor/recovery/code/provider/RecoveryCodeFactorProvider.java]: error log removed.

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/RootProvider.java (hunk 4 above)
Comment (ashraf706) [reply gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/RootProvider.java:516]: Recovery code added to cspHandler

**Comment by ashraf706** [comment]
> In addition of the two comments, here is 2 test cases on which I wondering if we should do something.
> 
> * First case
>   1 - enable a factor (SMS or OTP) on application and recovery code
>   2 - login a user and register the factor
>   3 - validate the factor challenge
>   4 - revory code page is displayed.
>   5 - do not click on the button but close the browser (or close the session)
>   6 - login the user again and validate the factor challenge
>   7 - the recovery code generation isn't done as it has been done during the first connection but as the session has been closed, but as we didn't click on the button can we consider that the user copied the code ?
> * Second case:
> 
> 1 - enable a factor (SMS or OTP) on application but NOT the recoevry code 2 - login a user and register the factor 3 - validate the factor challenge 4 - logout the user 5 - enable the recovery code on the app 6 - login the user 7 - validate the factor challenge 8 - recovery code are not displayed. (I expected recovery code page with my codes as the option has been enabled)

Fixed both use cases.

**Review by leleueri — APPROVED** [review]

**Review by tcompiegne — CHANGES_REQUESTED** [review]

File: gravitee-am-ui/src/app/domain/components/forms/form/dialog/form-info.component.html [hunk 6]
```diff
@@ -480,6 +480,40 @@ <h1>Access error</h1>
...
+                      </li>
+                  </ul>
+              </div>
+              <div>
+                  <button id="continue" th:onclick="(
+                  function(){
+                            let returnUrl = /*[[${returnURL}]]*/ '';
+                            window.location = returnUrl;
+                          }
+                  )();
```
Comment (tcompiegne) [inline gravitee-am-ui/src/app/domain/components/forms/form/dialog/form-info.component.html:507]: Move to the form version

**Review by tcompiegne — CHANGES_REQUESTED** [review]

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/resources/webroot/views/mfa_challenge.html [hunk 7]
```diff
@@ -60,6 +60,8 @@
                     </div>
                     <input type="hidden" th:name="${_csrf.parameterName}" th:value="${_csrf.token}"/>
                     <input type="hidden" th:name="factorId" th:value="${factor.id}"/>
+                    <input th:if="${factor.factorType.type == 'RECOVERY CODE'}" type="hidden" th:name="recoveryCodefactorId" th:value="${factor.id}"/>
```
Comment (tcompiegne) [inline gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/resources/webroot/views/mfa_challenge.html:63]: Not a huge fan of this if ..., we have already the factorId one line above.

**Review by tcompiegne — CHANGES_REQUESTED** [review]

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/resources/webroot/views/mfa_challenge.html (hunk 7 above)
Comment (tcompiegne) [reply gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/resources/webroot/views/mfa_challenge.html:63]: Seems to not be used anywhere

File: gravitee-am-ui/src/app/domain/components/forms/form/dialog/form-info.component.html (hunk 6 above)
Comment (ashraf706) [reply gravitee-am-ui/src/app/domain/components/forms/form/dialog/form-info.component.html]: done

File: gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/resources/webroot/views/mfa_challenge.html (hunk 7 above)
Comment (ashraf706) [reply gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/resources/webroot/views/mfa_challenge.html]: removed as it is not used

**Review by tcompiegne — APPROVED** [review]

**Pull request merged**